### PR TITLE
Implement Gaussian density with special covariance factorization and related

### LINKED ISSUE #61: Implement evaluation of a Gaussian density in case of special factorization of the covariance matrix
This issue is for the implementation of a utility function to evaluate a multivariate Gaussian density in those cases in which the covariance matrix can be written in the form `S = UV + R` with `R` an invertible block diagonal matrix.

This is useful when the covariance matrix of the measurement model is such that
- the evaluation of its **inverse** and
- the evaluation of its **determinant**,

that are required to evaluate the Gaussian density, are computationally demanding because `rows(S)` is a large number. However, if `cols(U) << rows(U)`, then it possible to use the  [Sherman-Morrison formula](https://en.wikipedia.org/wiki/Sherman%E2%80%93Morrison_formula) and the fact that `R` is block diagonal to reduce the problem to the inversion (and the evaluation of the determinant) of a matrix of size `cols(U) x cols(U)`

Once this is implemented, we will be able to implement `bfl::SUKFCorrection::getLikelihood`.

### PR DESCRIPTION
**Pull request by xEnVrE**
Within this PR:
- implemented a utility function, `utils::multivariate_gaussian_density_UVR`, for evaluation of a multivariate Gaussian probability density in case of a covariance of the form `S = UV + R` with `R` a block diagonal matrix;
> This method is useful when `cols(U) << rows(U) = N`, where `N` is the size of the input to the Gaussian density, since it avoids inverting a `rows(U) x rows(U)` matrix and evaluating the determinant of the same matrix. Instead, by using the `Sherman-Morrison-Woodbury` identity, it inverts a matrix having size `cols(U) x cols(U)`.
As an example, in the context of `UKF` filtering with an additive measurement model, the predicted measurement covariance can be written as `S = UV + R = YY^{T} + R` with `Y` a matrix containing the predicted measurements sigma points, shifted w.r.t to the predicted mean and scaled using the square root of the unscented transform weights. In this case, `rows(Y)` is the number of measurements while `cols(Y)` is equal to `2 x n + 1` where `n` is the size of the state vector. If the number of measurements is much greater than `2 x n + 1`, then the condition `cols(U) << rows(U)` is verified.
- implemented method `SUKFCorrection::getLikelihood` using `utils::multivariate_gaussian_density_UVR`;
- added test `test_Gaussian_High_Dimensional` testing method `utils::multivariate_gaussian_density_UVR`;
- updated `CHANGELOG.md`.

This closes #61

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,8 @@
 - Implemented method ParticleSet::resize().
 - Implemented evaluation of a multivariate Gaussian probability density function in method utils::multivariate_gaussian_density.
 - Implemented evaluation of the logarithm of a multivariate Gaussian probability density function in method utils::multivariate_gaussian_log_density.
+- Implemented evaluation of a multivariate Gaussian probability density function for high dimensional inputs in method utils::multivariate_gaussian_density_high_dim.
```
Comment (claudiofantacci) [inline CHANGELOG.md]: Please use `` delimiters to function, classes and everything related to the library. In this case it should be `utils::multivariate_gaussian_density_high_dim`. We'll update the rest of the document prior to the next release.

File: CHANGELOG.md [hunk 2]
```diff
@@ -14,6 +14,8 @@
...
+- Implemented evaluation of the logarithm of a multivariate Gaussian probability density function for high dimenstional inputs in method utils::multivariate_gaussian_log_density_high_dim.
```
Comment (claudiofantacci) [inline CHANGELOG.md]: Same as above.

File: CHANGELOG.md [hunk 3]
```diff
@@ -28,6 +30,7 @@
 - Implemented method KFCorrection::getLikelihood().
 - Implemented method UKFPrediction::getStateModel().
 - Implemented method UKFCorrection::getLikelihood().
+- Implemented method SUKFCorrection::getLikelihood().
```
Comment (claudiofantacci) [inline CHANGELOG.md]: Same as above.

File: CHANGELOG.md [hunk 4]
```diff
@@ -42,6 +45,7 @@
 - Add mean extraction and logging in test_SIS to simplify inspection of the algorithm outcome.
 - Reduce number of particles in test_SIS to reduce testing computation time in Debug.
 - Add testUPF_MAP testing MAP (maximum a posteriori) estimate extraction within a UPF particle filter.
+- Add test_Gaussian_High_Dimensional testing the method utils::multivariate_gaussian_density_high_dim().
```
Comment (claudiofantacci) [inline CHANGELOG.md]: Same as above.

File: src/BayesFilters/include/BayesFilters/utils.h [hunk 5]
```diff
@@ -83,6 +83,102 @@ Eigen::VectorXd multivariate_gaussian_log_density(const Eigen::MatrixBase<Derive
...
+ *
+ * This version is much faster than the standard gaussian logarithm density evaluation
+ * if U.cols() << U.rows().
+ *
+ * @param input Input representing the argument of the function as a vector or matrix.
+ * @param mean The mean of the associated Gaussian distribution as a vector.
+ * @param U The U factor within the covariance matrix S = UV + R of the associated Gaussian distribution as a matrix.
+ * @param V The V factor within the covariance matrix S = UV + R of the associated Gaussian distribution as a matrix.
+ * @param R The R summand within the covariance matrix S = UV + R of the associated Gaussian distribuion as a matrix.
+ *        The matrix R has size M * (N * M) and consists in the concatenation of all the diagonal blocks R_{i}.
```
Comment (claudiofantacci) [inline src/BayesFilters/include/BayesFilters/utils.h]: Please indent this and the following line with the previous one starting from `The R summand ...` 😄

File: src/BayesFilters/include/BayesFilters/utils.h [hunk 6]
```diff
@@ -99,6 +195,34 @@ Eigen::VectorXd multivariate_gaussian_density(const Eigen::MatrixBase<Derived>&
...
+ * dimension M such that N * M is the size of the input for some positive integer N.
+ *
+ * This version is much faster than the standard gaussian density evaluation
+ * if U.cols() << U.rows().
+ *
+ * @param input Input representing the argument of the function as a vector or matrix.
+ * @param mean The mean of the associated Gaussian distribution as a vector.
+ * @param U The U factor within the covariance matrix S = UV + R of the associated Gaussian distribution as a matrix.
+ * @param V The V factor within the covariance matrix S = UV + R of the associated Gaussian distribution as a matrix.
+ * @param R The R summand within the covariance matrix S = UV + R of the associated Gaussian distribuion as a matrix.
```
Comment (claudiofantacci) [inline src/BayesFilters/include/BayesFilters/utils.h:213]: Same as for the previous function.

File: src/BayesFilters/include/BayesFilters/utils.h [hunk 7]
```diff
@@ -83,6 +83,102 @@ Eigen::VectorXd multivariate_gaussian_log_density(const Eigen::MatrixBase<Derive
...
+ *        The matrix R has size M * (N * M) and consists in the concatenation of all the diagonal blocks R_{i}.
+ *        If the diagonal blocks are all equal, it is possible to provide a single M * M block matrix.
+ *
+ * @return The value of the logarithm of the density function evaluated on the input data as a vector.
+ */
+template<typename Derived>
+Eigen::VectorXd multivariate_gaussian_log_density_high_dim(const Eigen::MatrixBase<Derived>& input, const Eigen::Ref<const Eigen::VectorXd>& mean, const Eigen::Ref<const Eigen::MatrixXd>& U, const Eigen::Ref<const Eigen::MatrixXd>& V, const Eigen::Ref<const Eigen::MatrixXd>& R)
```
Comment (claudiofantacci) [inline src/BayesFilters/include/BayesFilters/utils.h]: I don't quite get the name of this function. Wouldn't it be more informative if it would be something like `multivariate_gaussian_log_density_UVR`? This function does not account, in general, high dimensional matrices, but ones that can be decomposed in a particular form. I would write that "form factor" information in the function name 😄

File: test/CMakeLists.txt [hunk 8]
```diff
@@ -11,6 +11,7 @@ option(TEST_LOG_TO_FILE "Write test logs to file" OFF)
 
 add_subdirectory(test_DirectionalStatisticsUtils)
 add_subdirectory(test_Gaussian)
+add_subdirectory(test_Gaussian_High_Dimensional)
```
Comment (claudiofantacci) [inline test/CMakeLists.txt]: Same here with this name. I would rename it accordingly 😄

**Review by claudiofantacci — CHANGES_REQUESTED** [review]
> This method is useful when `rows(U) << cols(U) = N`, where `N` is the size of the input to the Gaussian density, since it avoids inverting a `rows(U) x rows(U)` matrix and evaluating the determinant of the same matrix. Instead, by using the `Sherman-Morrison-Woodbury` identity, it inverts a matrix having size `cols(U) x cols(U)`;

The columns of `U` (whose cardinality is `N`) are the measurements, am I right?
We would like to invert a smaller matrix, possibly of `row(U)` dimension rather than a bigger `col(U)` one, isn't it?
Also, in the documentation of `multivariate_gaussian_log_density_high_dim` you wrote
> This version is much faster than the standard gaussian logarithm density evaluation if U.cols() << U.rows().
I'm a bit confuesed 🙃 Could you please explain and or correct your statements?

This is a great PR! 🎉 Thanks for the effort! I have just some minor correction here and there, nothing related to the code. It would be useful if you drop down the mathematical form of the implemented function so that I can double check it. Otherwise I'll trust you on that 😆

**Comment by xEnVrE** [comment]
>I'm a bit confused :upside_down_face: Could you please explain and or correct your statements?

Sorry, there was a typo in the description of the PR :laughing: I corrected it!

> It would be useful if you drop down the mathematical form of the implemented function so that I can double check it. Otherwise, I'll trust you on that :laughing:

I wrote a test to verify the output of the method since I was afraid there might be errors in the code.

File: CHANGELOG.md (hunk 1 above)
Comment (xEnVrE) [reply CHANGELOG.md]: Ok :+1:

File: CHANGELOG.md (hunk 2 above)
Comment (xEnVrE) [reply CHANGELOG.md]: Ok :+1:

File: src/BayesFilters/include/BayesFilters/utils.h (hunk 5 above)
Comment (xEnVrE) [reply src/BayesFilters/include/BayesFilters/utils.h]: Ok :+1:

File: src/BayesFilters/include/BayesFilters/utils.h (hunk 7 above)
Comment (xEnVrE) [reply src/BayesFilters/include/BayesFilters/utils.h]: I agree, being high dimensional is just a particular case. Do you want me to change the title of the PR and the associated issue?

Comment (claudiofantacci) [reply src/BayesFilters/include/BayesFilters/utils.h]: Yes please! 👍

**Review by claudiofantacci — APPROVED** [review]
LGTM 🎉 Thanks!
